presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -421,7 +422,7 @@ private static URI getBufferLocation(RemoteTask remoteTask, OutputBufferId rootB
             if (isInternalSystemConnector(connectorId)) {
                 connectorId = null;
             }
-            NodeSelector nodeSelector = nodeScheduler.createNodeSelector(connectorId);
+            NodeSelector nodeSelector = nodeScheduler.createNodeSelector(connectorId, maxTasksPerStage);
```

Comment:
The interface now is really weird.

`nodeScheduler.createNodeSelector` now acceps the `maxTasksPerStage`. And then the `nodeSelector.selectRandomNodes` accepts the `maxTasksPerStage`

I wonder if passing the limit to the `NodeSelector#computeAssignments` is a lesser evil?

CC: @wenleix

Reply:
I think changing `NodeSelector#computeAssignments`is a more evil cuz it will require to change TopologyAwareNodeSelector too.

Reply:
> I think changing NodeSelector#computeAssignmentsis a more evil

Let me elaborate a little bit more on this:

Creating the `NodeSelector` with the limit (that in fact the `maxTasksPerStage` is) is quite confusing.

The `NodeSelector` has other methods, that accept the limit as parameters. Thus it is not clear what limit will be applied, say in the `NodeSelector#selectRandomNodes`? The `maxTasksPerStage` limit? Or the `selectRandomNodes(limit..`)? Or `min()` / `max()`?

> it will require to change TopologyAwareNodeSelector too.

Currently it is simply ignored anyway. So i don't see a big difference.

Anyhow, i don't feel particularly strong about this. Feel free to leave it as is if you do.

Reply:
@arhimondr , @cemcayiroglu : If we are OK to have `maxTasksPerStage` when creating the `NodeSelector`, do we want to just pre-compute a limited set of nodes when `NodeSelector` is created? 

Thus the only reason why need to consult `nodeManager` during `computeAssignments` is when the query is submitted at cluster initialization time, which is not a common case.

Reply:
> Thus the only reason why need to consult nodeManager during computeAssignments is when the query is submitted at cluster initialization time, which is not a common case.

I'm not sure how important that in practice, but i think it is better to preserve the existing behaviour.

Reply:
Looks like this part of the code is super messy. nodescheduler and nodeselector is calling each other. no one knows who is doing what.  I am creating an issue to address this. please look: https://en.wikipedia.org/wiki/Spaghetti_code

Reply:
https://github.com/prestodb/presto/issues/13553

Reply:
https://github.com/prestodb/presto/issues/13553